[c10d] Adopt allgather_into_tensor_coalesced for NCCL.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103086

This is done by adding c10d::_allgather_into_tensor_coalesced wrapper.

kumpera committed:
[c10d] Adopt allgather_into_tensor_coalesced for NCCL.

This is done by adding c10d::_allgather_into_tensor_coalesced wrapper.

[ghstack-poisoned]

kumpera committed:
Update on "[c10d] Adopt allgather_into_tensor_coalesced for NCCL."


This is done by adding c10d::_allgather_into_tensor_coalesced wrapper.

[ghstack-poisoned]

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4211,6 +4211,65 @@ def _get_group_tag(pg: ProcessGroup) -> str:
+        return
+
+    input_tensor_list = [
+        t if not t.is_complex() else torch.view_as_real(t) for t in input_tensor_list
```
if we're supporting complex, do we need to add tests?

kumpera replied:
This method is gone and now we use coalescing manager instead.

rohan-varma commented on torch/testing/_internal/distributed/multi_threaded_pg.py:
```diff
@@ -341,6 +341,12 @@ def _reduce_scatter_base(self, output_tensor, input_tensor, opts=AllgatherOption
         tensor_list = list(torch.chunk(input_tensor, self._world_size))
         return self.reduce_scatter([output_tensor], [tensor_list], opts)
 
+    def allgather_into_tensor_coalesced(self, output_tensor_list, input_tensor_list):
```
should it be a private API in MTPG since its a private API for NCCL?

kumpera replied:
This is the name of the method in the PG api, we could fix it to be private.

rohan-varma reviewed (approved):
LGTM, some q's

kumpera committed:
Update on "[c10d] Adopt allgather_into_tensor_coalesced for NCCL."


This is done by adding c10d::_allgather_into_tensor_coalesced wrapper.

[ghstack-poisoned]

wanchaol commented on torch/distributed/_functional_collectives.py:
```diff
@@ -704,20 +704,21 @@ def _all_gather_into_tensor_coalesced_fallback(output_tensors, input_tensors, gr
+        with c10d._coalescing_manager(group=group, async_ops=True) as cm:
+            for in_t, out_t in zip(input_tensors, output_tensors):
+                dist.all_gather_into_tensor(out_t, in_t, group=group, async_op=True)
+        return cm
```
hmmm I think we should return the work/asyncCollectiveTensor list instead of `cm` for the nccl path?

kumpera replied:
Actually no, when using async collectives with _CoalescingManager, we need to call wait on it as the individual collectives won't return a work object.

kumpera committed:
Update on "[c10d] Adopt allgather_into_tensor_coalesced for NCCL."


This is done by adding c10d::_allgather_into_tensor_coalesced wrapper.

[ghstack-poisoned]

kumpera committed:
Update on "[c10d] Adopt allgather_into_tensor_coalesced for NCCL."


This is done by adding c10d::_allgather_into_tensor_coalesced wrapper.

[ghstack-poisoned]

kumpera committed:
Update on "[c10d] Adopt allgather_into_tensor_coalesced for NCCL."


This is done by adding c10d::_allgather_into_tensor_coalesced wrapper.

[ghstack-poisoned]

kumpera commented:
@pytorchmergebot merge
